Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs

```diff
@@ -1,46 +1,276 @@
+        }
+
+        /// <summary>
+        /// Disables the analyzer when the user closes the UI
+        /// </summary>
+        private void OnDisabledMessage(EntityUid uid, GasAnalyzerComponent component, GasAnalyzerDisableMessage message)
+        {
+            if (message.Session.AttachedEntity is not {Valid: true} user)
+                return;
+            // The UI should've already been closed but no harm in making sure?
+            DisableAnalyzer(uid, component, user);
+        }
+
+        private void OpenUserInterface(EntityUid user, GasAnalyzerComponent component)
+        {
+            if (!TryComp<ActorComponent>(user, out var actor))
+                return;
+
+            _userInterface.TryOpen(component.Owner, GasAnalyzerUiKey.Key, actor.PlayerSession);
+        }
+
+        /// <summary>
+        /// Fetches fresh data for the analyzer. Should only be called by Update or when the user requests an update via refresh button
+        /// </summary>
+        private bool UpdateAnalyzer(EntityUid uid, GasAnalyzerComponent? component = null)
+        {
+            if (!Resolve(uid, ref component))
+                return false;
+
+            // check if the user has walked away from what they scanned
+            var userPos = EntityManager.GetComponent<TransformComponent>(component.User).Coordinates;
+            if (component.LastPosition.HasValue)
+            {
+                // Check if position is out of range => don't update and disable
+                if (!component.LastPosition.Value.InRange(EntityManager, userPos, SharedInteractionSystem.InteractionRange))
+                {
+                    DisableAnalyzer(uid, component, component.User);
+                    return false;
+                }
+            }
+
+            var gasMixList = new List<GasMixEntry>();
 
-            if (TryComp(args.User, out ActorComponent? actor))
+            if (component.Target != null)
             {
-                component.OpenInterface(actor.PlayerSession, args.ClickLocation);
+                // gas analyzed was used on an entity, try to request gas data
+                var ev = new GasAnalyzerScanEvent();
+                RaiseLocalEvent((EntityUid)component.Target, ev, false);
+
+                if (ev.GasMixtures != null)
+                {
+                    foreach (var mixes in ev.GasMixtures)
+                    {
+                        if(mixes.Value != null)
+                            gasMixList.Add(new GasMixEntry(mixes.Key, mixes.Value.Pressure, mixes.Value.Temperature, GenerateGasEntryArray(mixes.Value)));
+                    }
+                }
+                else
+                {
+                    // TODO: HACKY SOLUTION UNTIL I FIGURE OUT HOW TO HANDLE PIPES VIA EVENT
+                    // note though this will pick up any atmospherics devices that don't add a handler for the scan event
+                    // But they really should add that in themselves
```

Comment:
this seems like a decent catch-all solution to me; as mentioned above maybe it should just default to this for all stuff? i suppose there could be a point to be made about not including specific gas mixtures if the system doesn't want them to be included, but you could solve that by reintroducing `GasAnalyzable` in some form

Reply:
When I was first inquiring on this I was advised that an event system would be the way to go instead of using component references. I was able to do it for everything but pipes, hence this work around. But maybe I misunderstood what they meant?

Reply:
oh--by component references i specifically meant the `[ComponentReference]` attribute + interfaces on the components, getting the nodecontainer is fine

Reply:
Ohh gotcha, yeah I definitely misunderstood that. I can simplify it a bit in that case as any device that has a NodeContainer then would work fine. It would just leave gas tanks, canisters (it will report the Port not the mix), portables like the scrubber that would still have to rely on the event